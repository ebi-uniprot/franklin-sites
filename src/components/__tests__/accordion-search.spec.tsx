import { screen, render, fireEvent } from '@testing-library/react';

import AccordionSearch, {
  filterAccordionData,
  getLeafKeys,
} from '../accordion-search';

jest.mock('lodash-es', () => ({
  debounce: (fn: unknown) => fn,
}));

const props = {
  placeholder: 'Filter',
  onSelect: jest.fn(),
  selected: [],
  accordionData: [
    {
      label: 'Gene',
      id: '1',
      items: [
        {
          label: 'BRCA1',
          id: '1-1',
        },
        {
          label: 'BRCA2',
          id: '1-2',
        },
        {
          label: 'TP53',
          id: '1-3',
        },
      ],
    },
    {
      label: 'Organelle',
      id: '2',
      items: [
        {
          label: 'Ribosome',
          id: '2-1',
        },
        {
          label: 'Nucleus',
          id: '2-2',
        },
      ],
    },
  ],
};

describe('AccordionSearch', () => {
  it('should render', () => {
    const { asFragment } = render(<AccordionSearch {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('filterAccordionData', () => {
    it('should return filtered data with matching query', () => {
      const filteredAccordionData = filterAccordionData(
        props.accordionData,
        'nucleus'
      );
      expect(filteredAccordionData).toHaveLength(1);
      expect(filteredAccordionData[0].items?.[0].id).toBe('2-2');
    });

    it('should return no data with nonmatching query', () => {
      const filteredAccordionData = filterAccordionData(
        props.accordionData,
        'zap'
      );
      expect(filteredAccordionData).toHaveLength(0);
    });
  });

  it('should find correct number of items when input is entered', async () => {
    render(<AccordionSearch {...props} />);
    const input = screen.getByTestId('search-input');
    fireEvent.change(input, { target: { value: 'Nucleus' } });
    const checkboxes = await screen.findAllByRole('checkbox');
    expect(checkboxes).toHaveLength(1);
  });

  it('should call onSelect when item clicked', () => {
    render(<AccordionSearch {...props} />);
    const checkboxes = screen.getAllByRole('checkbox');
    if (checkboxes.length) {
      fireEvent.click(checkboxes[0]);
    }
    expect(props.onSelect).toHaveBeenCalled();
  });

  describe('getLeafKeys', () => {
    it('should get all of the leaf keys', () => {
      expect(getLeafKeys(props.accordionData)).toEqual([
        '1-1',
        '1-2',
        '1-3',
        '2-1',
        '2-2',
      ]);
    });
  });
});
