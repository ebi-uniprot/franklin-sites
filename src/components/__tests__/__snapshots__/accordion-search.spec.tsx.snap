// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionSearch should render 1`] = `
<DocumentFragment>
  <div
    style="width: 18rem;"
  >
    <span
      class="search-input"
    >
      <input
        data-testid="search-input"
        placeholder="Filter"
        type="text"
        value=""
      />
      <span
        class="search-input__suffix"
        data-testid="search-input-suffix"
        role="presentation"
      >
        <test-file-stub
          height="14"
          width="14"
        />
      </span>
    </span>
  </div>
  <div
    class="accordion-group accordion-search"
  >
    <div
      class="accordion"
    >
      <div
        class="accordion__title"
        data-testid="accordion-title"
        role="button"
        tabindex="0"
      >
        <div
          class="accordion__title__text"
        >
          Gene
        </div>
        <div
          class="accordion__title__side"
        >
          <test-file-stub
            classname="accordion__title__side__chevron"
            height="16"
            width="16"
          />
        </div>
      </div>
      <div
        class="accordion__content accordion__content--hide-content"
        data-testid="accordion-content"
      >
        <ul
          class="no-bullet accordion-search__list"
        >
          <li
            class="accordion-search__list__item"
            data-testid="accordion-search-list-item"
          >
            <label
              for="checkbox-1-1"
            >
              <input
                class="accordion-search__list__item-checkbox"
                id="checkbox-1-1"
                type="checkbox"
              />
              BRCA1
            </label>
          </li>
          <li
            class="accordion-search__list__item"
            data-testid="accordion-search-list-item"
          >
            <label
              for="checkbox-1-2"
            >
              <input
                class="accordion-search__list__item-checkbox"
                id="checkbox-1-2"
                type="checkbox"
              />
              BRCA2
            </label>
          </li>
          <li
            class="accordion-search__list__item"
            data-testid="accordion-search-list-item"
          >
            <label
              for="checkbox-1-3"
            >
              <input
                class="accordion-search__list__item-checkbox"
                id="checkbox-1-3"
                type="checkbox"
              />
              TP53
            </label>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="accordion"
    >
      <div
        class="accordion__title"
        data-testid="accordion-title"
        role="button"
        tabindex="0"
      >
        <div
          class="accordion__title__text"
        >
          Organelle
        </div>
        <div
          class="accordion__title__side"
        >
          <test-file-stub
            classname="accordion__title__side__chevron"
            height="16"
            width="16"
          />
        </div>
      </div>
      <div
        class="accordion__content accordion__content--hide-content"
        data-testid="accordion-content"
      >
        <ul
          class="no-bullet accordion-search__list"
        >
          <li
            class="accordion-search__list__item"
            data-testid="accordion-search-list-item"
          >
            <label
              for="checkbox-2-1"
            >
              <input
                class="accordion-search__list__item-checkbox"
                id="checkbox-2-1"
                type="checkbox"
              />
              Ribosome
            </label>
          </li>
          <li
            class="accordion-search__list__item"
            data-testid="accordion-search-list-item"
          >
            <label
              for="checkbox-2-2"
            >
              <input
                class="accordion-search__list__item-checkbox"
                id="checkbox-2-2"
                type="checkbox"
              />
              Nucleus
            </label>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
